chem_tensorflow_sparse.py

```diff
@@ -122,33 +120,73 @@ def compute_final_node_representations(self) -> tf.Tensor:
-                        for adjacency_matrix in adjacency_matrices:
-                            incoming_messages_per_type = tf.sparse_tensor_dense_matmul(adjacency_matrix,
-                                                                                       node_states_per_layer[-1])  # v x D
-                            incoming_messages.extend([incoming_messages_per_type])
-
-                        # Pass incoming messages through linear layer:
-                        incoming_messages = tf.concat(incoming_messages, axis=1)  # v x [2 *] edge_types
-                        messages_passed = tf.matmul(incoming_messages,
-                                                    self.weights['edge_weights'][layer_idx])  # v x D
+                        for edge_type_idx, adjacency_list_for_edge_type in enumerate(self.placeholders['adjacency_lists']):
+                            edge_sources, edge_targets = adjacency_list_for_edge_type[:, 0], adjacency_list_for_edge_type[:, 1]
+                            edge_source_states = tf.nn.embedding_lookup(params=node_states_per_layer[-1],
+                                                                        ids=edge_sources)  # Shape [E, D]
+                            all_messages_for_edge_type = tf.matmul(edge_source_states,
+                                                                   self.weights['edge_weights'][layer_idx][edge_type_idx])  # Shape [E, D]
+                            messages.append(all_messages_for_edge_type)
+                            message_source_states.append(edge_source_states)
+                            message_targets.append(edge_targets)
+                            message_edge_types.append(tf.ones_like(edge_targets, dtype=tf.int32) * edge_type_idx)
+
+                        messages = tf.concat(messages, axis=0)  # Shape [M, D]
+                        message_targets = tf.concat(message_targets, axis=0)  # Shape [M]
+
+                        if self.params['use_propagation_attention']:
+                            message_source_states = tf.concat(message_source_states, axis=0)  # Shape [M, D]
+                            message_target_states = tf.nn.embedding_lookup(params=node_states_per_layer[-1],
+                                                                           ids=message_targets)  # Shape [M, D]
+                            message_attention_scores = tf.einsum('mi,mi->m', message_source_states, message_target_states)  # Shape [M]
+
+                            message_edge_types = tf.concat(message_edge_types, axis=0)  # Shape [M]
+                            message_edge_type_factors = tf.nn.embedding_lookup(params=self.weights['edge_type_attention_weights'][layer_idx],
+                                                                               ids=message_edge_types)  # Shape [M]
+
+                            message_attention_scores = message_attention_scores * message_edge_type_factors
```

Comment:
Let's test both an additive and a multiplicative version?

Reply:
Should I make this configurable via a hyperparam?

Reply:
Why not...